daceml/autodiff/backward_pass_generator.py

```diff
@@ -0,0 +1,1270 @@
+
+log = logging.getLogger(__name__)
+
+
+def _strings_to_symbols(strings: typing.Set[str]) -> typing.Set[sp.Symbol]:
+    return {sp.symbols(string) for string in strings}
+
+
+def _symbols_to_strings(symbs: typing.Set[sp.Symbol]) -> typing.Set[str]:
+    return {str(symb) for symb in symbs}
+
+
+def generate_grad_connector_names(
+        existing_connectors: typing.Set[str],
+        forward_connector_names: typing.List[str]) -> typing.Dict[str, str]:
+    """ Choose connector names for the gradients of all forward connectors.
+
+        :param existing_connectors: existing connectors on the node.
+        :param forward_connector_names: the list of connectors to generate names for.
+        :returns: a mapping from entries in ``forward_connector_names`` to names for those entries.
+    """
+
+    # copy
+    existing_connectors = set(existing_connectors)
+
+    names = {}
+    for n in forward_connector_names:
+        result = find_str_not_in_set(existing_connectors, n + "_gradient")
+        names[n] = result
+        existing_connectors.add(result)
+
+    return names
+
+
+def is_initialization_state(state: SDFGState) -> bool:
+    """ Check if state is an initialization state, i.e. it initializes one or more arrays with zero values
+    """
+    for n in state.data_nodes():
+        if len(state.out_edges(n)) > 0:
+            return False
+    return True
+
+
+def code_to_exprs(code: str, inputs: typing.Set[str],
+                  outputs: typing.Set[str]) -> typing.Dict[str, sp.Expr]:
+    """Convert a python string to a set of (simplified) symbolic sympy expressions. Currently, this
+    supports only code consisting of assignment statements.
+    :param code: the code to convert
+    :param code: the inputs (i.e. the defined variables) for the code
+    :param code: the outputs to generate simplified expressions for
+    :return: map from outputs to symbolic expressions
+    """
+
+    inputs = list(inputs)
+    outputs = list(outputs)
+
+    code_fn = """
+def symbolic_execution({}):
+    # define functions from cmath.h
+    from sympy import exp, log
+    def log2(x):
+        return log(x, 2)
+    def log10(x):
+        return log(x, 10)
+    from sympy import sin, cos, tan, asin, acos, atan, sinh, cosh, tanh, asinh, acosh, atanh
+    from sympy import sin, cos, tan, asin, acos, atan, sinh, cosh, tanh, asinh, acosh, atanh
+    from sympy import Pow as pow, sqrt
+    from sympy import sign, floor, ceiling as ceil, Abs as abs, Abs as fabs
+    from sympy import Max as max, Min as min
+    from sympy import Max as fmax, Min as fmin
+{}
+    return {}
+    """
+    code_fn = code_fn.format(
+        ", ".join(inputs),
+        "\n".join("    " + line.strip() for line in code.split("\n")),
+        ", ".join(outputs),
+    )
+
+    try:
+        exec(code_fn)
+
+        # no idea why, but simply calling symbolic_execution doesn't work
+        results = vars()["symbolic_execution"](
+            *[sp.symbols(inp) for inp in inputs])
```

Comment:
hacky and not always correct. you should just pass in two extra args to exec, for globals and locals. The resulting function will be defined in the second arg (globals)